gateway/src/lib.rs

```diff
@@ -147,16 +209,121 @@ pub mod pallet {
+                }
+            };
+            (dispatch_weight, dispatch_class)
+        })]
+        pub fn execute(
+            origin: OriginFor<T>,
+            proof: Vec<u8>,
+            chain_id: u32,
+            command_ids: Vec<H256>,
+            commands: Vec<String>,
+            calls: Vec<<T as Config>::RuntimeCall>,
+        ) -> DispatchResultWithPostInfo {
+            let _ = ensure_signed(origin)?;
+
+            // TODO: Once XCM is enabled this check might not make sense
+            ensure!(chain_id == T::ChainId::get(), Error::<T>::WrongChainId);
+
+            ensure!(
+                calls.len() == command_ids.len(),
+                Error::<T>::CommandIdsLengthMismatch
+            );
+
+            let payload = Self::abi_encode_batch_params(
+                chain_id.clone(),
+                command_ids.clone(),
+                commands.clone(),
+                calls.clone(),
+            );
+            let mut is_active_operators =
+                Self::validate_proof(H256::from(sp_core::keccak_256(&payload)), &proof)?;
+
+            let gateway_origin = RawOrigin::Signed(Self::account_id());
+            // Code simplified taken from https://github.com/paritytech/substrate/blob/ee316317b85b2f65fc022b27bbfefcd42b6560ae/frame/utility/src/lib.rs#L440
+            let calls_len = calls.len();
+            ensure!(
+                calls_len <= Self::batched_calls_limit() as usize,
+                Error::<T>::TooManyCalls
+            );
+
+            // Track the actual weight of each of the batch calls.
+            let mut weight = Weight::zero();
+            // Track failed dispatch occur.
+            let mut has_error: bool = false;
+            for (idx, call) in calls.into_iter().enumerate() {
+                if <CommandExecuted<T>>::get(command_ids[idx]) {
+                    continue;
+                }
+
+                if let Some(Call::transfer_operatorship { .. }) = call.is_sub_type() {
+                    if !is_active_operators {
+                        continue;
+                    }
+                    is_active_operators = false;
+                // TODO: Do we need the approve flow or trigger final execution instead
+                // else if { Some(Call::approve_contract_call { .. }) = call.is_sub_type() }
+                } else {
+                    // Do not execute if call not allowed, just skip
+                    continue;
+                }
+
+                let info = call.get_dispatch_info();
+                <CommandExecuted<T>>::set(command_ids[idx], true);
+                // TODO: Should we honor the original sender? if so how to do that
+                // If origin is root, don't apply any dispatch filters; root can call anything.
+                let result = call.dispatch(gateway_origin.clone().into());
+                // Add the weight of this call.
+                weight = weight.saturating_add(extract_actual_weight(&result, &info));
```

Comment:
We only use `info` at line 294 and only once so I suggest dropping this line and inline its usage: 

```suggestion
                <CommandExecuted<T>>::set(command_ids[idx], true);
                // TODO: Should we honor the original sender? if so how to do that
                // If origin is root, don't apply any dispatch filters; root can call anything.
                let result = call.dispatch(gateway_origin.clone().into());
                // Add the weight of this call.
                weight = weight.saturating_add(extract_actual_weight(&result, &call.get_dispatch_info()));
```